rmf_site_format/src/workcell.rs

```diff
@@ -289,6 +344,110 @@ impl Workcell {
         serde_json::ser::to_string_pretty(self)
     }
 
+    pub fn to_urdf(&self) -> Result<urdf_rs::Robot, WorkcellToUrdfError> {
+        let workcell = self.clone();
+
+        let visuals_and_parents: Vec<(urdf_rs::Visual, _)> = workcell
+            .visuals
+            .into_iter()
+            .map(|v| {
+                let visual = v.1.bundle;
+                let visual = urdf_rs::Visual {
+                    name: Some(visual.name),
+                    origin: visual.pose.into(),
+                    geometry: visual.geometry.into(),
+                    material: None,
+                };
+                (visual, v.1.parent)
+            })
+            .collect();
+
+        let collisions_and_parents: Vec<(urdf_rs::Collision, _)> = workcell
+            .collisions
+            .into_iter()
+            .map(|c| {
+                let collision = c.1.bundle;
+                let collision = urdf_rs::Collision {
+                    name: Some(collision.name),
+                    origin: collision.pose.into(),
+                    geometry: collision.geometry.into(),
+                };
+                (collision, c.1.parent)
+            })
+            .collect();
+
+        let links: Result<Vec<_>, WorkcellToUrdfError> = workcell
+            .frames
+            .into_iter()
+            .map(|f| {
+                let frame = f.1.bundle;
+
+                let name = match frame.name {
+                    Some(name) => name.0,
+                    None => format!("frame_{}", f.0),
+                };
+                let visual: Vec<Visual> = visuals_and_parents
+                    .iter()
+                    .filter(|(_, parent)| parent == &f.1.parent)
+                    .map(|(visual, _)| visual.clone())
+                    .collect();
+                let collision: Vec<Collision> = collisions_and_parents
+                    .iter()
+                    .filter(|(_, parent)| parent == &f.1.parent)
+                    .map(|(collision, _)| collision.clone())
+                    .collect();
```

Comment:
I wonder if there is a more efficient way to do this filtering.
What is happening here is a nested iteration where, for each link, we iterate through all the visuals / collisions to check if any of them is a parent of the link.
What if you did a HashMap where either:
* Instead of a `Vec<Link>` you have a `HashMap<u32, Link>` then populate it once when iterating through all the links with their ID and data or
* You keep a `Vec<Link>` but build in parallel a `HashMap<u32, &mut Link>` (don't quote me on this not sure it would actually work! It would in principle avoid doing the conversion from HashMap -> Vec when exporting).

Then you only iterate through visuals / collisions once and for each item you find which link to append it to by doing a hash map lookup.
This should flatten the nested loop and make the export a lot faster

Reply:
I went with a hashmap of vectors for the links and collisions in 41e04ff.

I think this is better as we're not iterating over everything, every time. We just do it once now.